pkg/apitype/core.go

```diff
@@ -198,6 +234,50 @@ type ResourceV2 struct {
 	Provider string `json:"provider,omitempty" yaml:"provider,omitempty"`
 }
 
+// ResourceV3 is the third version of the Resource API type. It absorbs a few breaking changes:
+//   1. It adds a map from input property names to the dependencies that affect that input property. This is used to
+//      improve the precision of delete-before-create operations.
+//   2. It adds a new boolean field, `PendingReplacement`, that marks resources that have been deleted as part of a
+//      delete-before-create operation but have not yet been recreated.
+//
+// Migrating from ResourceV2 to ResourceV3 involves:
+//   1. Populating the map from input property names to dependencies by assuming that every dependency listed in
+//      `Dependencies` affects every input property.
+type ResourceV3 struct {
+	// URN uniquely identifying this resource.
+	URN resource.URN `json:"urn" yaml:"urn"`
+	// Custom is true when it is managed by a plugin.
+	Custom bool `json:"custom" yaml:"custom"`
+	// Delete is true when the resource should be deleted during the next update.
+	Delete bool `json:"delete,omitempty" yaml:"delete,omitempty"`
+	// ID is the provider-assigned resource, if any, for custom resources.
+	ID resource.ID `json:"id,omitempty" yaml:"id,omitempty"`
+	// Type is the resource's full type token.
+	Type tokens.Type `json:"type" yaml:"type"`
+	// Inputs are the input properties supplied to the provider.
+	Inputs map[string]interface{} `json:"inputs,omitempty" yaml:"inputs,omitempty"`
+	// Outputs are the output properties returned by the provider after provisioning.
+	Outputs map[string]interface{} `json:"outputs,omitempty" yaml:"outputs,omitempty"`
+	// Parent is an optional parent URN if this resource is a child of it.
+	Parent resource.URN `json:"parent,omitempty" yaml:"parent,omitempty"`
+	// Protect is set to true when this resource is "protected" and may not be deleted.
+	Protect bool `json:"protect,omitempty" yaml:"protect,omitempty"`
+	// External is set to true when the lifecycle of this resource is not managed by Pulumi.
+	External bool `json:"external,omitempty" yaml:"external,omitempty"`
+	// Dependencies contains the dependency edges to other resources that this depends on.
+	Dependencies []resource.URN `json:"dependencies,omitempty" yaml:"dependencies,omitempty"`
+	// InitErrors is the set of errors encountered in the process of initializing resource (i.e.,
+	// during create or update).
+	InitErrors []string `json:"initErrors,omitempty" yaml:"initErrors,omitempty"`
+	// Provider is a reference to the provider that is associated with this resource.
+	Provider string `json:"provider,omitempty" yaml:"provider,omitempty"`
+	// PropertyDependencies maps from an input property name to the set of resources that property depends on.
+	PropertyDependencies map[resource.PropertyKey][]resource.URN `json:"propertyDependencies,omitempty" yaml:"property_dependencies,omitempty"`
```

Comment:
why the `snake_case` here in the yaml case? (though, surely we don't need yaml tags on the checkpoint apitype... 😆 )